tez-runtime-library/src/main/java/org/apache/tez/runtime/library/common/sort/impl/IFile.java

```diff
@@ -71,6 +73,223 @@
+          Math.max(getBaseCacheSize(), cacheStream.available());
+      this.fileCodec = codec;
+    }
+
+    /**
+     * For basic cache size checks: header + checksum + EOF marker
+     *
+     * @return size of the base cache needed
+     */
+    static int getBaseCacheSize() {
+      return (checksumSize + (2 * WritableUtils.getVIntSize(EOF_MARKER))
+          + HEADER.length);
+    }
+
+    /**
+     * Create in mem stream. In it is too small, adjust it's size
+     *
+     * @param size
+     * @return in memory stream
+     */
+    public static BoundedByteArrayOutputStream createBoundedBuffer(int size) {
+      int resize = Math.max(getBaseCacheSize(), size);
+      return new BoundedByteArrayOutputStream(resize);
+    }
+
+    /**
+     * Flip over from memory to file based writer.
+     *
+     * 1. Content format: HEADER + real data + CHECKSUM. Checksum is for real
+     * data.
+     * 2. Before flipping, close checksum stream, so that checksum is written
+     * out.
+     * 3. Create relevant file based writer.
+     * 4. Write header and then real data.
+     *
+     * @throws IOException
+     */
+    private void resetToFileBasedWriter() throws IOException {
+      // Close out stream, so that data checksums are written.
+      // Buf contents = HEADER + real data + CHECKSUM
+      this.out.close();
+
+      // Get the buffer which contains data in memory
+      BoundedByteArrayOutputStream bout =
+          (BoundedByteArrayOutputStream) this.rawOut.getWrappedStream();
+
+      // Create new file based writer
+      if (outputPath == null) {
+        outputPath = taskOutput.getOutputFileForWrite();
+      }
+      LOG.info("Creating file " + outputPath);
+      FSDataOutputStream newRawOut = fs.create(outputPath);
+      this.rawOut = newRawOut;
+
+      this.checksumOut = new IFileOutputStream(this.rawOut);
+      if (fileCodec != null) {
+        this.compressor = CodecPool.getCompressor(fileCodec);
+        if (this.compressor != null) {
+          this.compressor.reset();
+          this.compressOutput = true;
+          this.compressedOut = fileCodec.createOutputStream(checksumOut, compressor);
+          this.out = new FSDataOutputStream(this.compressedOut, null);
+        } else {
+          LOG.warn("Could not obtain compressor from CodecPool");
+          this.out = new FSDataOutputStream(checksumOut, null);
+        }
+      } else {
+        this.out = new FSDataOutputStream(checksumOut, null);
+      }
+
+      // Write header to file
+      headerWritten = false;
+      writeHeader(newRawOut);
+
+      // write real data
+      int sPos = HEADER.length;
+      int len = (bout.size() - checksumSize - HEADER.length);
+      this.out.write(bout.getBuffer(), sPos, len);
+
+      bufferFull = true;
+      bout.reset();
+    }
+
+    @Override
+    public void close() throws IOException {
+      try {
+        super.close();
+      } finally {
+        if (this.rawOut != null) {
```

Comment:
Why is this required?

Reply:
Otherwise rawOut wouldn't be closed causing issues premature EOF exception. Actually this override can be avoided if we set `ownOutputStream=true`. In this case, `Writer::close` itself would close the stream. Will fix it in the next commit.